sql/parser/parser.go

```diff
@@ -19,11 +19,17 @@ type Parser struct {
 	orderedParams int
 	namedParams   int
 	buf           *bytes.Buffer
+	functions     expr.Functions
 }
 
 // NewParser returns a new instance of Parser.
 func NewParser(r io.Reader) *Parser {
-	return &Parser{s: scanner.NewBufScanner(r)}
+	return &Parser{s: scanner.NewBufScanner(r), functions: expr.NewFunctions()}
+}
+
+func (p *Parser) WithFunctions(functions expr.Functions) *Parser {
+	p.functions = functions
```

Comment:
I think `WithFunctions` should return a shallow copy.

```go
p2 := *p
p2.functions = functions
return &p2
```

Reply:
I think we shouldn't use the `With*` pattern here because as @tie suggested it's better to return shallow copies when doing so, which would force us to duplicate buffers, perhaps reset readers, etc.

What we want is to create a parser with some specific functions. How about we just make the constructor more flexible?

```go
func NewParser(r io.Reader, opts *Options) *Parser {
 ...
}

type Options struct {
  Functions expr.Functions
  // in the future we could add more options here
}
```

`nil` options means using defaults. We are already doing so in other packages